### PR TITLE
New module semodule

### PR DESCRIPTION
**Pull request by bandit145**
##### SUMMARY
Adds an semodule Ansible module for compiling and applying .te (Type enforcement files) as discussed in https://github.com/ansible/ansible/issues/27349

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
semodule

##### ADDITIONAL INFORMATION
Most of the logic for controlling this module is written in the action, this seemed like a good way to cut down on needing to copy the .te file to the target host if not needed.

```paste below
.te file:
module ansible-semodule-test 1.0;

require {
	type proc_net_t;
	type sshd_t;
	class file { open read };
}

allow sshd_t proc_net_t:file { open read};

- name: install test module
  semodule:
    src: test.te

TASK [semodule : install test module] *********************************
changed: [testhost] => {"changed": true, "name": "ansible-semodule-test", "version": "1.0"}

TASK [semodule : debug] ********************************************************
ok: [testhost] => {
    "test1_mod_install": {
        "changed": true, 
        "failed": false, 
        "name": "ansible-semodule-test", 
        "version": "1.0"
    }
}
```

### CODE REVIEW

File: plugins/action/semodule.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            if module_installed:
+                check_res = self._check_mode(True, policy_def['name'], result)
+                if check_res:
+                    return check_res
+                self._remove_module(policy_def['name'])
+                result['changed'] = True
+            else:
+                result['changed'] = False
+        # clean up temp path
+        self._remove_tmp_path(self._connection._shell.tmpdir)
```
Comment (felixfontein) [inline plugins/action/semodule.py:94]: Why not do that in a `try`/`finally:`?

Comment (bandit145) [reply plugins/action/semodule.py]: Good point, not sure why I didn't do that. Should be good now.

File: plugins/modules/system/semodule.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        rc, stdout, stderr = module.run_command(['which', req])
+        if rc != 0:
+            module.fail_json(msg='missing the {name} utility which is required to run this module'.format(name=req))
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            src=dict(type='path', required=True),
+            force=dict(type='bool', required=False),
```
Comment (felixfontein) [inline plugins/modules/system/semodule.py:117]:
```suggestion
            force=dict(type='bool', default=False),
```

File: plugins/modules/system/semodule.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    type: bool
+requirements:
+  - policycoreutils
+  - checkpolicy
+author:
+  - Philip Bove (@bandit145)
+'''
+
+EXAMPLES = '''
+# Install policy module from type enforcement file
```
Comment (felixfontein) [inline plugins/modules/system/semodule.py:45]: Please use `name:` of a task instead of adding a comment.

**Comment by bachradsusi** [comment]
This module has the same name as "semodule" tool but it's purpose is different. `/usr/sbin/semodule` can't build SELinux modules, but it provides the following functionality: install, enable, disable, remove, export (in pp HLL or directly in CIL), list. It also use priorities which can be used to override default modules. So using the same name could be confusing. 
I'd expect that *semodule* module follows `semodule` tools behaviour: install/remove *pp* or *cil* modules with default or defined priority, enable/disable modules, ...

As I wrote in https://github.com/ansible/ansible/issues/27349#issuecomment-582411694 sources of SELinux modules consist of 3 files - *module.te*, *module.fc*, *module.if*.  *.te* files usually use M4 macros so you need to process them first. Without *.fc* files you are not able to define a new file context. E.g. how would you deploy https://github.com/USBGuard/usbguard-selinux using this module?

I don't think it makes sense to build SELinux modules only from ".te" files directly on nodes. SELinux modules are usually deployed as *pp* files, or *CIL* files so i's better to have them prebuild in *pp* HLL or *CIL* language.

**Comment by bandit145** [comment]
@bachradsusi Sorry that I missed your comment on the original issue. I agree with you that this is a poor name for the module and it seems of limited use now that you have explained the typical deployment method for SELinux modules. 

We do use the proposed method quite a bit at work but I'll go ahead and close this due to the seemingly limited use and in retrospect obvious issue with the name.

Thanks

**Comment by felixfontein** [comment]
Some points for discussion:
1. Is it possible to extend the module to cover the more general case? Or does the design prevent this?
2. Is there a use-case for the current module that warrants including it, albeit with a different name? (So that the name `semodule` can be used for something which fits better?)

If not, @bandit145 you can always publish the module in your own collection and publish it on Ansible Galaxy; that way people who need exactly this functionality can easily install and use it without having to copy files to the correct places.

**Comment by redgryphon** [comment]
I found this PR while looking for something that does exactly what @bandit145 proposed.

The code could be a little limited for the whole spectrum of options offered by the `semodule` tool but it is not a reason to stop working thus I wouldn't have closed this PR. I would have modified it as a WIP because there is a solid base to work on.
What do you think, @bandit145?

**Pull request closed without merging**